Improve handling of dynamic discovery. (#3596)
ros2/ros2_documentation: source/Tutorials/Advanced/Improved-Dynamic-Discovery.rst

**Time:** 15 minutes

.. contents:: Table of Contents
   :depth: 2
   :local:

Overview
--------

By default, ROS 2 will attempt to find all nodes on all hosts on the same subnet automatically.
However, the following options are available to control the ROS 2 discovery range.

Configuration Parameters
------------------------

* ``ROS_AUTOMATIC_DISCOVERY_RANGE``: controls how far ROS nodes will try to discover each other.

   Valid options are:

   * ``SUBNET`` is the default, and for DDS based middleware it means it will discover any node reachable via multicast.
   * ``LOCALHOST`` means a node will only try to discover other nodes on the same machine.
   * ``OFF`` means the node won't discover any other nodes, even on the same machine.
   * ``SYSTEM_DEFAULT`` means "don't change any discovery settings".

* ``ROS_STATIC_PEERS``: is a semicolon (``;``) separated list of addresses that ROS should try to discover nodes on.
  This allows connecting to nodes on specific machines (as long as their discovery range is not set to ``OFF``).

The combination of these two environment variables for local and remote nodes will enable and control the ROS 2 communication discovery range.

> The following tables complete the all possible combination for those two variables and how the discovery range behaves.

Reviewer: ```suggestion
The following tables highlight the discovery range behavior for possible conbinations.
```

     - ``X``
     - ``O``
     - ``O``
     - ``X``
     - ``O``
     - ``O``

.. list-table:: Node A and B running in the different hosts
   :widths: 20 20 20 20 20 20 20 20 20
   :header-rows: 1

   * - Different hosts
     -
     -
     - Node B setting
     -
     -
     -
     -
     -
   * -
     -
     -
     - No static peer
     -
     -
     - With static peer
     -
     -
   * -
     -
     -
     - Off
     - Localhost
     - Subnet
     - Off
     - Localhost
     - Subnet
   * - Node A setting
     - No static peer
     - Off
     - ``X``
     - ``X``
     - ``X``
     - ``X``
     - ``X``
     - ``X``
   * -
     -
     - Localhost
     - ``X``
     - ``X``
     - ``X``
     - ``X``
     - ``O``
     - ``O``
   * -
     -
     - Subnet
     - ``X``
     - ``X``
     - ``O``
     - ``X``
     - ``O``
     - ``O``
   * -
     - With static peer
     - Off
     - ``X``
     - ``X``
     - ``X``
     - ``X``
     - ``X``
     - ``X``
   * -
     -
     - Localhost
     - ``X``
     - ``O``
     - ``O``
     - ``X``
     - ``O``
     - ``O``
   * -
     -
     - Subnet
     - ``X``
     - ``O``
     - ``O``
     - ``X``
     - ``O``
     - ``O``

Examples
--------

> For example, the following commands will limit the ROS 2 communication only with local host and specific peers:

Reviewer: ```suggestion
For example, the following commands will limit the ROS 2 communication only with localhost and specific peers:
```

> .. redirect-from::
> 
>     Improved-Dynamic-Discovery
>     Tutorials/Improved-Dynamic-Discovery/Improved-Dynamic-Discovery
> 
> .. _ImprovedDynamicDiscovery:
> 

Reviewer 2: Since this is a new document, we don't need any of this preamble (there are no previous links here).  So just remove this whole bit:

```suggestion
```

Author: thanks, `redirect-from` has been removed.

Improved Dynamic Discovery
==========================

**Goal:** This tutorial will show how to use the improved dynamic discovery configuration.

**Tutorial level:** Advanced

**Time:** 15 minutes

.. contents:: Table of Contents
   :depth: 2
   :local:

Overview
--------

> Iron Irwini or later, improved dynamic discovery options are available to control the ROS 2 communication range.
> By default, ROS 2 communication is not limited, so nodes on other hosts that are on the same subnet will be found automatically.
> Improved dynamic discovery provides the following environmental variable to control ROS 2 communication discovery.

Reviewer 2: ```suggestion
By default, ROS 2 will attempt to find all nodes on all hosts on the same subnet automatically.
However, the following options are available to control the ROS 2 discovery range.
```
